addons/stock/models/stock_picking.py

```diff
@@ -1038,48 +1041,48 @@ def check_destinations(self):
-                            rounding_method='UP')
-                        done_to_keep = ml.qty_done
-                        new_move_line = ml.copy(
-                            default={'product_uom_qty': 0, 'qty_done': ml.qty_done})
-                        ml.write({'product_uom_qty': quantity_left_todo, 'qty_done': 0.0})
-                        new_move_line.write({'product_uom_qty': done_to_keep})
-                        move_lines_to_pack |= new_move_line
-
-                package_level = self.env['stock.package_level'].create({
-                    'package_id': package.id,
-                    'picking_id': pick.id,
-                    'location_id': False,
-                    'location_dest_id': move_line_ids.mapped('location_dest_id').id,
-                    'move_line_ids': [(6, 0, move_lines_to_pack.ids)]
-                })
-                move_lines_to_pack.write({
-                    'result_package_id': package.id,
-                })
-            else:
-                raise UserError(_('You must first set the quantity you will put in the pack.'))
+        for pick in self:
+            move_lines_to_pack = self.env['stock.move.line']
+            package = self.env['stock.quant.package'].create({})
+            for ml in move_line_ids:
+                if float_compare(ml.qty_done, ml.product_uom_qty,
+                                 precision_rounding=ml.product_uom_id.rounding) >= 0:
+                    move_lines_to_pack |= ml
+                else:
+                    quantity_left_todo = float_round(
+                        ml.product_uom_qty - ml.qty_done,
+                        precision_rounding=ml.product_uom_id.rounding,
+                        rounding_method='UP')
+                    done_to_keep = ml.qty_done
+                    new_move_line = ml.copy(
+                        default={'product_uom_qty': 0, 'qty_done': ml.qty_done})
+                    ml.write({'product_uom_qty': quantity_left_todo, 'qty_done': 0.0})
+                    new_move_line.write({'product_uom_qty': done_to_keep})
+                    move_lines_to_pack |= new_move_line
+            package_level = self.env['stock.package_level'].create({
+                'package_id': package.id,
+                'picking_id': pick.id,
+                'location_id': False,
+                'location_dest_id': move_line_ids.mapped('location_dest_id').id,
+                'move_line_ids': [(6, 0, move_lines_to_pack.ids)]
+            })
+            move_lines_to_pack.write({
+                'result_package_id': package.id,
+            })
         return package
 
     def put_in_pack(self):
-        res = self.check_destinations()
-        if res.get('type'):
-            return res
-        return self._put_in_pack()
+        for picking in self.filtered(lambda p: p.state not in ('done', 'cancel')):
+            move_line_ids = picking.move_line_ids.filtered(lambda o: o.qty_done > 0 and not o.result_package_id)
```

Comment:
this filtered on the move line is written 3 times in this pr, i guess it would make sense to write it in an helper and use float_compare?